File: packages/api-models-py/skyportal_py_models/followup_requests.py
"""Request and response models for ``/api/followup_request``."""

from __future__ import annotations

import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from skyportal_py_models.allocations import Allocation
from skyportal_py_models.groups import Group
from skyportal_py_models.users import User


class FacilityTransaction(BaseModel):
    """A serialized exchange with a facility (upstream ``FacilityTransaction``).

    ``followup_request`` and ``observation_plan_request`` are the parent
    rows; they stay ``dict`` to avoid a circular import back into this
    module and into :mod:`skyportal_py_models.observation_plans`.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime.datetime | None = None
    modified: datetime.datetime | None = None
    request: dict[str, Any] | None = None
    response: dict[str, Any] | None = None
    followup_request_id: int | None = None
    observation_plan_request_id: int | None = None
    initiator_id: int | None = None
    initiator: User | None = None
    followup_request: dict[str, Any] | None = None
    observation_plan_request: dict[str, Any] | None = None


class FacilityTransactionRequest(BaseModel):
    """A queued facility call (upstream ``FacilityTransactionRequest``).

    ``followup_request`` and ``observation_plan_request`` are the parent
    rows; they stay ``dict`` to avoid a circular import back into this
    module and into :mod:`skyportal_py_models.observation_plans`.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime.datetime | None = None
    modified: datetime.datetime | None = None
    last_query: datetime.datetime | None = None
    method: str | None = None
    endpoint: str | None = None
    data: dict[str, Any] | None = None
    params: dict[str, Any] | None = None
    headers: dict[str, Any] | None = None
    status: str | None = None
    followup_request_id: int | None = None
    observation_plan_request_id: int | None = None
    initiator_id: int | None = None
    initiator: User | None = None
    followup_request: dict[str, Any] | None = None
    observation_plan_request: dict[str, Any] | None = None


class FollowupRequestWatcher(BaseModel):
    """A user watching a follow-up request (upstream ``FollowupRequestUser``)."""

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime.datetime | None = None
    modified: datetime.datetime | None = None
    followuprequest_id: int | None = None
    user_id: int | None = None


class FollowupRequest(BaseModel):
    """A follow-up observation request (upstream ``FollowupRequest``).

    ``obj`` stays ``dict`` because typing it as
    :class:`skyportal_py_models.sources.Source` would create an import cycle;
    the same applies to ``photometry``, ``photometric_series`` and
    ``spectra``, which all point back at the requesting object.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime.datetime | None = None
    modified: datetime.datetime | None = None
    obj_id: str | None = None
    allocation_id: int | None = None
    requester_id: int | None = None
    last_modified_by_id: int | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    status: str | None = None
    comment: str | None = None
    obj: dict[str, Any] | None = None
    allocation: Allocation | None = None
    requester: User | None = None
    last_modified_by: User | None = None
    target_groups: list[Group] = Field(default_factory=list)
    watchers: list[FollowupRequestWatcher] = Field(default_factory=list)
    transactions: list[FacilityTransaction] = Field(default_factory=list)
    transaction_requests: list[FacilityTransactionRequest] = Field(default_factory=list)
    photometry: list[dict[str, Any]] = Field(default_factory=list)
    photometric_series: list[dict[str, Any]] = Field(default_factory=list)
    spectra: list[dict[str, Any]] = Field(default_factory=list)
    rise_time_utc: str | list[str] | None = None
    set_time_utc: str | list[str] | None = None


class FollowupRequestsPage(BaseModel):
    """One page of results from a follow-up requests query."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    followup_requests: list[FollowupRequest] = Field(default_factory=list)
    total_matches: int = Field(alias="totalMatches", default=0)
    page_number: int = Field(alias="pageNumber", default=1)
    num_per_page: int = Field(alias="numPerPage", default=100)


class FollowupRequestPost(BaseModel):
    """Payload for submitting a follow-up request."""

    model_config = ConfigDict(extra="forbid")

    obj_id: str
    allocation_id: int
    payload: dict[str, Any]
    target_group_ids: list[int] | None = None


class FollowupRequestPostResponse(BaseModel):
    """Result of submitting a follow-up request."""

    model_config = ConfigDict(extra="forbid")

    id: int


class DefaultFollowupRequest(BaseModel):
    """A default follow-up request (upstream ``DefaultFollowupRequest``)."""

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime.datetime | None = None
    modified: datetime.datetime | None = None
    requester_id: int | None = None
    allocation_id: int | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    default_followup_name: str | None = None
    source_filter: dict[str, Any] | str | None = None
    constraints: dict[str, Any] | None = None
    priority_order: str | None = None
    validity_days: int | None = None
    comment: str | None = None
    implements_update: bool | None = None
    allocation: Allocation | None = None
    requester: User | None = None
    target_groups: list[Group] = Field(default_factory=list)


class DefaultFollowupRequestPost(BaseModel):
    """Payload for creating a default follow-up request."""

    model_config = ConfigDict(extra="forbid")

    allocation_id: int
    payload: dict[str, Any]
    default_followup_name: str
    source_filter: dict[str, Any]
    target_group_ids: list[int] | None = None
    comment: str | None = None
    implements_update: bool | None = None
    priority_order: str | None = None
    validity_days: int | None = None
    radius: float | None = None
    not_if_duplicates: bool | None = None
    source_group_ids: list[int] | None = None
    ignore_source_group_ids: list[int] | None = None
    not_if_classified: bool | None = None
    not_if_spectra_exist: bool | None = None
    not_if_tns_classified: bool | None = None
    not_if_tns_reported: float | None = None
    not_if_assignment_exists: bool | None = None
    ignore_allocation_ids: list[int] | None = None


class DefaultFollowupRequestPostResponse(BaseModel):
    """Result of creating a default follow-up request."""

    model_config = ConfigDict(extra="forbid")

    id: int


class PhotometryRequestStatus(BaseModel):
    """Status of a follow-up request after a photometry retrieval."""

    model_config = ConfigDict(extra="forbid")

    id: int
    request_status: str | None = None

File: packages/api-models-ts/src/FollowupRequests.ts
/**
 * Request and response models for `/api/followup_request`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Allocations from "./Allocations.ts";
import * as Groups from "./Groups.ts";
import * as Schemas from "./Schemas.ts";
import * as Users from "./Users.ts";

/**
 * A serialized exchange with a facility (upstream `FacilityTransaction`).
 *
 * `followup_request` and `observation_plan_request` are the parent rows; they
 * stay free-form to avoid a circular import back into this module and into
 * {@link skyportal-js/ObservationPlans}.
 *
 * @since 1.0.0
 * @category Models
 */
export const FacilityTransaction = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        request: Schemas.nullish(Schemas.JsonObject),
        response: Schemas.nullish(Schemas.JsonObject),
        followup_request_id: Schemas.NullishInteger,
        observation_plan_request_id: Schemas.NullishInteger,
        initiator_id: Schemas.NullishInteger,
        initiator: Schemas.nullish(Users.User),
        followup_request: Schemas.nullish(Schemas.JsonObject),
        observation_plan_request: Schemas.nullish(Schemas.JsonObject),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type FacilityTransaction = v.InferOutput<typeof FacilityTransaction>;

/**
 * A queued facility call (upstream `FacilityTransactionRequest`).
 *
 * @since 1.0.0
 * @category Models
 */
export const FacilityTransactionRequest = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        last_query: Schemas.NullishTimestamp,
        method: Schemas.NullishString,
        endpoint: Schemas.NullishString,
        data: Schemas.nullish(Schemas.JsonObject),
        params: Schemas.nullish(Schemas.JsonObject),
        headers: Schemas.nullish(Schemas.JsonObject),
        status: Schemas.NullishString,
        followup_request_id: Schemas.NullishInteger,
        observation_plan_request_id: Schemas.NullishInteger,
        initiator_id: Schemas.NullishInteger,
        initiator: Schemas.nullish(Users.User),
        followup_request: Schemas.nullish(Schemas.JsonObject),
        observation_plan_request: Schemas.nullish(Schemas.JsonObject),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type FacilityTransactionRequest = v.InferOutput<typeof FacilityTransactionRequest>;

/**
 * A user watching a follow-up request (upstream `FollowupRequestUser`).
 *
 * @since 1.0.0
 * @category Models
 */
export const FollowupRequestWatcher = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        followuprequest_id: Schemas.NullishInteger,
        user_id: Schemas.NullishInteger,
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type FollowupRequestWatcher = v.InferOutput<typeof FollowupRequestWatcher>;

/**
 * The fields of a {@link FollowupRequest}.
 *
 * @since 1.0.0
 * @category Models
 */
export const FollowupRequestEntries = {
    id: Schemas.Integer,
    created_at: Schemas.NullishTimestamp,
    modified: Schemas.NullishTimestamp,
    obj_id: Schemas.NullishString,
    allocation_id: Schemas.NullishInteger,
    requester_id: Schemas.NullishInteger,
    last_modified_by_id: Schemas.NullishInteger,
    payload: v.optional(Schemas.JsonObject, () => ({})),
    status: Schemas.NullishString,
    comment: Schemas.NullishString,
    obj: Schemas.nullish(Schemas.JsonObject),
    allocation: Schemas.nullish(Allocations.Allocation),
    requester: Schemas.nullish(Users.User),
    last_modified_by: Schemas.nullish(Users.User),
    target_groups: Schemas.list(Groups.Group),
    watchers: Schemas.list(FollowupRequestWatcher),
    transactions: Schemas.list(FacilityTransaction),
    transaction_requests: Schemas.list(FacilityTransactionRequest),
    photometry: Schemas.list(Schemas.JsonObject),
    photometric_series: Schemas.list(Schemas.JsonObject),
    spectra: Schemas.list(Schemas.JsonObject),
    rise_time_utc: Schemas.nullish(v.union([v.string(), v.array(v.string())])),
    set_time_utc: Schemas.nullish(v.union([v.string(), v.array(v.string())])),
};

/**
 * A follow-up observation request (upstream `FollowupRequest`).
 *
 * `obj` stays free-form because typing it as
 * {@link skyportal-js/Sources!Source} would create an import cycle; the same
 * applies to `photometry`, `photometric_series` and `spectra`, which all point
 * back at the requesting object.
 *
 * @since 1.0.0
 * @category Models
 */
export const FollowupRequest = Schemas.model(v.strictObject(FollowupRequestEntries));

/**
 * @since 1.0.0
 * @category Models
 */
export type FollowupRequest = v.InferOutput<typeof FollowupRequest>;

/**
 * One page of results from a follow-up requests query.
 *
 * @since 1.0.0
 * @category Models
 */
export const FollowupRequestsPage = Schemas.model(
    v.strictObject({
        followup_requests: Schemas.list(FollowupRequest),
        totalMatches: v.optional(Schemas.Integer, 0),
        pageNumber: v.optional(Schemas.Integer, 1),
        numPerPage: v.optional(Schemas.Integer, 100),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type FollowupRequestsPage = v.InferOutput<typeof FollowupRequestsPage>;

/**
 * Payload for submitting a follow-up request.
 *
 * `payload` holds the instrument-specific request parameters; the allocation's
 * instrument API defines its schema. If `target_group_ids` is omitted, the
 * server applies its default visibility to the results.
 *
 * @since 1.0.0
 * @category Models
 */
export interface FollowupRequestPost {
    readonly obj_id: string;
    readonly allocation_id: number;
    readonly payload: Record<string, unknown>;
    readonly target_group_ids?: ReadonlyArray<number> | undefined;
}

/**
 * Result of submitting a follow-up request.
 *
 * @since 1.0.0
 * @category Models
 */
export const FollowupRequestPostResponse = Schemas.model(v.strictObject({ id: Schemas.Integer }));

/**
 * @since 1.0.0
 * @category Models
 */
export type FollowupRequestPostResponse = v.InferOutput<typeof FollowupRequestPostResponse>;

/**
 * A default follow-up request (upstream `DefaultFollowupRequest`).
 *
 * @since 1.0.0
 * @category Models
 */
export const DefaultFollowupRequest = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        requester_id: Schemas.NullishInteger,
        allocation_id: Schemas.NullishInteger,
        payload: v.optional(Schemas.JsonObject, () => ({})),
        default_followup_name: Schemas.NullishString,
        source_filter: Schemas.nullish(v.union([Schemas.JsonObject, v.string()])),
        constraints: Schemas.nullish(Schemas.JsonObject),
        priority_order: Schemas.NullishString,
        validity_days: Schemas.NullishInteger,
        comment: Schemas.NullishString,
        implements_update: Schemas.NullishBoolean,
        allocation: Schemas.nullish(Allocations.Allocation),
        requester: Schemas.nullish(Users.User),
        target_groups: Schemas.list(Groups.Group),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type DefaultFollowupRequest = v.InferOutput<typeof DefaultFollowupRequest>;

/**
 * Payload for creating a default follow-up request.
 *
 * `payload` holds the instrument-specific request parameters and must not
 * contain `start_date` or `end_date` (the server fills the real window when
 * the request fires on source save). `source_filter` decides which saved
 * sources trigger the request and is required.
 *
 * @since 1.0.0
 * @category Models
 */
export interface DefaultFollowupRequestPost {
    readonly allocation_id: number;
    readonly payload: Record<string, unknown>;
    readonly default_followup_name: string;
    readonly source_filter: Record<string, unknown>;
    readonly target_group_ids?: ReadonlyArray<number> | undefined;
    readonly comment?: string | undefined;
    readonly implements_update?: boolean | undefined;
    readonly priority_order?: string | undefined;
    readonly validity_days?: number | undefined;
    readonly radius?: number | undefined;
    readonly not_if_duplicates?: boolean | undefined;
    readonly source_group_ids?: ReadonlyArray<number> | undefined;
    readonly ignore_source_group_ids?: ReadonlyArray<number> | undefined;
    readonly not_if_classified?: boolean | undefined;
    readonly not_if_spectra_exist?: boolean | undefined;
    readonly not_if_tns_classified?: boolean | undefined;
    readonly not_if_tns_reported?: number | undefined;
    readonly not_if_assignment_exists?: boolean | undefined;
    readonly ignore_allocation_ids?: ReadonlyArray<number> | undefined;
}

/**
 * Result of creating a default follow-up request.
 *
 * @since 1.0.0
 * @category Models
 */
export const DefaultFollowupRequestPostResponse = Schemas.model(v.strictObject({ id: Schemas.Integer }));

/**
 * @since 1.0.0
 * @category Models
 */
export type DefaultFollowupRequestPostResponse = v.InferOutput<typeof DefaultFollowupRequestPostResponse>;

/**
 * Status of a follow-up request after a photometry retrieval.
 *
 * @since 1.0.0
 * @category Models
 */
export const PhotometryRequestStatus = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        request_status: Schemas.NullishString,
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type PhotometryRequestStatus = v.InferOutput<typeof PhotometryRequestStatus>;
